src/Microsoft.DotNet.Build.Tasks/PackageFiles/publishtest.targets

```diff
@@ -97,52 +101,50 @@
       <TestWithoutNativeImages>true</TestWithoutNativeImages>
     </PropertyGroup>
 
-    <!-- Remove duplicates. Note that we mustn't just copy in sequence and let
-         the last one win that way because it will cause copies to occur on
-         every incremental build. -->
     <ItemGroup>
-      <_TestCopyLocalByFileName Include="@(TestCopyLocal->'%(RecursiveDir)%(FileName)%(Extension)')">
+      <!-- Not all platforms can use the .ni.dlls that come from packages.  If TestWithoutNativeImages is specified, we'll exclude them from copy generation.
+           If we end up needing this for any other sorts of filtering, we'll want to add a list of filtered extensions to be matched on EndsWith.  -->
+      <_TestCopyLocalByFileName Include="@(TestCopyLocal->'%(RecursiveDir)%(FileName)%(Extension)')"
+                                Condition="'$(TestWithoutNativeImages)' != 'true' Or !$([System.String]::Copy('%(Identity)').EndsWith('.ni.dll'))">
         <NugetPackageId>%(TestCopyLocal.NugetPackageId)</NugetPackageId>
+        <PackageRelativePath Condition="'%(TestCopyLocal.NugetPackageId)' != ''">$([System.String]::Copy('%(Identity)').Replace('$(PackagesDir)',''))</PackageRelativePath>
         <SourcePath>%(Identity)</SourcePath>
+        <UseAbsolutePath Condition="'$(TestWithLocalLibraries)'=='true' Or '$(TestWithLocalNativeLibraries)'=='true'">$([System.String]::Copy('%(Identity)').StartsWith('$(BinDir)'))</UseAbsolutePath>
       </_TestCopyLocalByFileName>
     </ItemGroup>
 
+    <!-- Remove duplicates. Note that we mustn't just copy in sequence and let
+         the last one win that way because it will cause copies to occur on
+         every incremental build. -->
     <RemoveDuplicatesWithLastOneWinsPolicy Inputs="@(_TestCopyLocalByFileName)">
       <Output TaskParameter="Filtered" ItemName="_TestCopyLocalByFileNameWithoutDuplicates" />
     </RemoveDuplicatesWithLastOneWinsPolicy>
 
-    <!-- Ideally, we'd have UseHardLinksIfPossible on by default because we copy tons of the
-         same files for every single test project here, but it doesn't currently work x-plat.
-         so we only turn it on by default for Windows builds for now. -->
+    <ComputeDestinationsForDependencies TestDependencies="@(_TestCopyLocalByFileNameWithoutDuplicates)">
+      <Output TaskParameter="TestDependenciesWithDestinations" ItemName="IncludedFileForRunnerScript" />
+    </ComputeDestinationsForDependencies>
+  </Target>
+
+  <!-- Remove items from IncludedFileForRunnerScript which are already in SourcesToCopyToTestDir.
```

Comment:
Right; the difference is that it uses the destination path for the comparison, which could include subdirectories.  Although using the Exclude attribute is cleaner, it would in this case require making two new item types which have the relative destination path as their ItemSpec, so I think it's preferable to use Remove based on the Condition that the destination path metadata for the IncludedFileForRunnerScript and SourceToCopyToTestDir are equal.